### ISSUE TITLE
Angular Element inside dynamic components gets destroyed twice.

### ISSUE THREAD

**Issue by Kartik1607** [angular/angular#29606]
# 🐞 bug report

### Affected Package
 The issue is caused by package @angular/elements

### Description

We are creating a plugin based dashboard where plugins are angular libraries which are injected dynamically. Libraries expose regular angular components which can be instantiated at run time. Also, libraries expose **angular elements** components which can be used by other libraries.

Let's say there are 2 libraries
1. Product Library : Exposes `ProductDetailComponent` as regular angular component. 
2. Cart Library : Exposes `AddToCartComponent` as angular element.

`ProductDetailComponent` refers to `AddToCartComponent` through HTML template.
```html
<div>
...
<add-to-cart></add-to-cart>
</div>
```

The Cart button renders perfectly as required. Problem occurs when we navigate away from the `ProductDetailComponent`.

```
e.html:1 ERROR TypeError: Cannot read property 'nativeNode' of null
    at removeDebugNodeFromIndex (core.js:18995)
    at DebugRenderer2.push../node_modules/@angular/core/fesm5/core.js.DebugRenderer2.destroyNode (core.js:24233)
    at destroyViewNodes (core.js:23461)
    at destroyView (core.js:23449)
    at callWithDebugContext (core.js:24176)
    at Object.debugDestroyView [as destroyView] (core.js:23884)
    at ViewRef_.push../node_modules/@angular/core/fesm5/core.js.ViewRef_.destroy (core.js:21710)
    at ComponentRef_.push../node_modules/@angular/core/fesm5/core.js.ComponentRef_.destroy (core.js:21546)
    at elements.js:236
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:423)
```
## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 7.3.6
Node: 9.11.1
OS: win32 x64
Angular: 7.2.10
... animations, common, compiler, compiler-cli, core, elements
... forms, language-service, platform-browser
... platform-browser-dynamic, router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.13.6
@angular-devkit/build-angular      0.13.6
@angular-devkit/build-ng-packagr   0.13.6
@angular-devkit/build-optimizer    0.13.6
@angular-devkit/build-webpack      0.13.6
@angular-devkit/core               7.3.6
@angular-devkit/schematics         7.3.6
@angular/cdk                       7.3.5
@angular/cli                       7.3.6
@angular/material                  7.3.5
@ngtools/json-schema               1.1.0
@ngtools/webpack                   7.3.6
@schematics/angular                7.3.6
@schematics/update                 0.13.6
ng-packagr                         4.7.1
rxjs                               6.3.3
typescript                         3.2.4
webpack                            4.29.0
</code></pre>

**Comment by juristr** [comment]
Oh interesting. I had the exact same issue a couple of days ago. @jasonaden I quickly created a repository that reproduces the issue: https://github.com/juristr/elements-bug-repro.
(You should be able to import it on Stackblitz here: [https://stackblitz.com/github/juristr/elements-bug-repro](https://stackblitz.com/github/juristr/elements-bug-repro))

If you start it and go to the `/home` route, there is an Angular Element in the HTML template. If you then click on the "Admin" like to change the route, you'll see the error message in the console of the devtools.

Interestingly, this apparently only happens if the Angular element is already embedded into the template. If you insert it dynamically like..

```
const el = document.createElement('test-element');
this.elRef.nativeElement.append(el);
```

...it doesn't happen.
I quickly looked at the stacktrace and it even seems like this is being just debug info. In fact, if you compile the example app and start the compiled one, the error doesn't happen.

**Comment by juristr** [comment]
So I debugged the issue quickly and the problem. Here's what I've found so far. 

The node gets inserted twice, passing in the `elements.ts:createElement(...)` function. Since it is present in the template, the first time it passes here:

![image](https://user-images.githubusercontent.com/542458/56021930-eb490c80-5d0a-11e9-81e8-2d5b630f83c8.png)

..and then once the Angular element gets instantiated, it again passes by here:

![image](https://user-images.githubusercontent.com/542458/56022067-35ca8900-5d0b-11e9-9ca3-1cd116e0c8e6.png)

..since in that scenario it is the root element.

Hence, the `indexDebugNode(...)` gets invoked two times as well:

![image](https://user-images.githubusercontent.com/542458/56022121-52ff5780-5d0b-11e9-9e0d-3132048adf0d.png)

The issue starts when we switch the view which gets destroyed. The according `removeDebugNodeFromIndex` gets called, and since it uses the elements `nativeNode` as key in the `Set` (which is the same instance), the first call removes it, and the 2nd fails.

**Comment by jaspreet1234567** [comment]
Hi @juristr  and @jasonaden ,

Any fix on this issue?

**Comment by juristr** [comment]
@jaspreet1234567 I have no update yet from the team. However, this bug is solved in v8, so if you can upgrade, that's the preferred way anyways (and would give you even more than just fixing this bug 😉 )

**Comment by Dr0drigues** [comment]
@juristr Hi,

When you say that this is solved in v8, do you mean on Angular 8? Because I just upgraded to Angular 8 and I'm still facing the issue :(

Is there any workaround or something to prevent this?

**Comment by juristr** [comment]
In fact, I though it was solved. Just had it today however. gonna have to dig deeper to see why it is still present

**Comment by jaspreet1234567** [comment]
> @juristr Hi,
> 
> When you say that this is solved in v8, do you mean on Angular 8? Because I just upgraded to Angular 8 and I'm still facing the issue :(
> 
> Is there any workaround or something to prevent this?

Same is here. One of my colleague tested this with angular v8 but it still has same issue.

**Comment by vasyliv** [comment]
+1 has the same bug.

**Comment by mrtarkhan** [comment]
the same after upgrading to Angular 8

**Comment by vasyliv** [comment]
here is a demo with error in Angular 8 
https://stackblitz.com/edit/angular-fpxg6m

**Comment by vasyliv** [comment]
is there any updates?

**Comment by mlakmal** [comment]
having same issue with child application using angular elements.

**Comment by tmburt04** [comment]
Any update on this?

**Comment by tmburt04** [comment]
I was able to alleviate the error being thrown by using my own 'ComponentNgElementStrategyFactory' and referencing it when defining my components.

**Custom component definition**
`  components.map(el => {
    const component = createCustomElement(el.component, {
      injector,
      strategyFactory: new XComponentNgElementStrategyFactory(
        el.component,
        injector
      )
    });
    customElements.get(el.selector) ||
      customElements.define(el.selector, component);
  });`

The strategy factory I'm using is almost exactly like the one provided by Angular, however, I modified the 'disconnect' method which seemed to solve the duplicated destruction.


**BEFORE**
`  /**
   * Schedules the component to be destroyed after some small delay in case the element is just
   * being moved across the DOM.
   */
  disconnect() {
    // Return if there is no componentRef or the component is already scheduled for destruction
    if (!this.componentRef || this.scheduledDestroyFn !== null) {
      return;
    }

    // Schedule the component to be destroyed after a small timeout in case it is being
    // moved elsewhere in the DOM
    this.scheduledDestroyFn = scheduler.schedule(() => {
      if (this.componentRef) {
        this.componentRef!.destroy();
        this.componentRef = null;
      }
    }, DESTROY_DELAY);
  }`

**AFTER**
` disconnect() {
    // Return if there is no componentRef or the component is already scheduled for destruction
    if (!this.componentRef || this.scheduledDestroyFn !== null) {
      return;
    }

    // Schedule the component to be destroyed after a small timeout in case it is being
    // moved elsewhere in the DOM
    this.scheduledDestroyFn = scheduler.schedule(() => {
      if (this.componentRef) {
        this.componentRef!.destroy();
        this.componentRef = null;
      }
    }, DESTROY_DELAY);
    this.scheduledDestroyFn();
  }`

Looking at the comments - 

>     // Schedule the component to be destroyed after a small timeout in case it is being
>     // moved elsewhere in the DOM

It seems like they expected the component to be destroyed, but they only defined the function, not calling it.

**Comment by DmitryEfimenko** [comment]
having this issue on Angular 8.2.12

**Comment by DaSchTour** [comment]
Will this be fixed with angular 9? I'm not sure if angular/elements is of any use with this bug in it.

**Comment by FredyValstrauss** [comment]
I'm facing this issue in Angular 8.2.14 when changing route from a component with SSRS embedded on it, to any other component.

**Comment by gkalpak** [comment]
This seems to only be an issue in ViewEngine (there is no error in Ivy - unless I have failed to reproduce it correctly):

- [Example with ViewEngine][1]
- [Example with Ivy][2]

<sub>If you are still seeing the error with Ivy, please provide an up-to-date reproduction 🙏</sub>

[1]: https://stackblitz.com/edit/ngelements-issue-29606-viewengine
[2]: https://stackblitz.com/edit/ngelements-issue-29606-ivy

**Comment by totszwai** [comment]
Is there ETA for this fix for Angular7?

**Comment by gkalpak** [comment]
Regarding https://github.com/angular/angular/issues/29606#issuecomment-538055772:

@tmburt04 (or anyone looking to implement https://github.com/angular/angular/issues/29606#issuecomment-538055772), the function returned by `scheduler.schedule()` (unfortunately named `scheduledDestroyFn`) is actually a function that can be used to cancel the scheduled fn. So, by calling `this.scheduledDestroyFn()` right away, you are immediately cancelling it.

This essentially makes the strategy's `disconnect()` method a no-op. I would be careful with such a change, because it can lead to unexpected behavior (e.g. memory leaks due to failing to destroy removed components).

##
@totszwai, v7 is not maintained any more, so this is not going to be fixed there. See our [Versioning and Releases][1] guide for more details. If all else fails (and you can't upgrade to the latest version and [Ivy][2], you can try @tmburt04's suggestion from https://github.com/angular/angular/issues/29606#issuecomment-538055772.)

[1]: https://angular.io/guide/releases#support-policy-and-schedule
[2]: https://v9.angular.io/guide/ivy

**Comment by totszwai** [comment]
@gkalpak What's the fix for post v7? Maybe I can port/patch it by altering suggestion from @tmburt04

**Comment by gkalpak** [comment]
@totszwai, this seems to be fixed in [Ivy][1] (our latest compiler+renderer), which was introduced as stable in v9. So, that is not really possible to port to v7, which still uses the old compiler/renderer, `ViewEngine`. In `ViewEngine`, this is still a problem (hence this issue is open).

[1]: https://v9.angular.io/guide/ivy

**Comment by TusharKanjariya** [comment]
Hello,
I am Using Angular Version 9. I am also faced this issue when i am navigate **one Module** to **other Modules** pages.

i was solved this Issue using

- Enable Ivy in **tsconfig.json** in `angularCompilerOptions` to set `"enableIvy":true`
- and adding in **src/polyfills.js** to add `import '@angular/localize/init';`

Hope, it will help.

**Comment by JoostK** [comment]
Closing as duplicate of #25525, which was fixed in Angular 11.2.10 in #41565.
